README.md
[![Build Status](https://travis-ci.org/mapbox/raster-tile-query.svg?branch=master)](https://travis-ci.org/mapbox/raster-tile-query)
raster-tile-query
=================

:warning: **This project is a proof of concept and not intended for production use.** :warning:

If you are looking to just get pixel values from an image, we recommend using [Node Mapnik](http://mapnik.org/documentation/node-mapnik/) and [sphericalmercator](https://github.com/mapbox/sphericalmercator) directly instead of this library. This is explained below.

To query raster tiles, such as [Mapbox Terrain RGB tiles](https://blog.mapbox.com/global-elevation-data-6689f1d0ba65), at particular longitudes and latitudes, you'll need to know the X, Y, Z, and pixel size of your image tile to be able to interpolate which relative pixel to get information from. XYZ info is generally found in the request URL of your tile like this:

```
https://api.mapbox.com/v4/mapbox.terrain-rgb/3/2/3.pngraw?access_token={your-api-token}
https://api.mapbox.com/v4/mapbox.terrain-rgb/{z}/{x}/{y}.pngraw?access_token={your-api-token}
```

The following main steps are taken to get relative pixel values:

* convert your lng/lat coordinates into pixel coordinates for your individual image.
* Determine the pixel location of your image with x,y,z and image pixel size.
* then subtract the two steps above to get relative pixel location for your image. If your latitude and longitude do not line up with the image, they will either be negative in number, or greater than the image size in pixels.
* Use an image reading library such as Node Mapnik or [get-pixels](https://www.npmjs.com/package/get-pixels) to retrieve pixel data

Here is an example script using Node Mapnik

```JavaScript
var fs = require('fs');
var mapnik = require('mapnik');
var SM = require('@mapbox/sphericalmercator');
var sm = new SM();

/**
 * Get the pixel values of an image from a longitude and latitude when proviiding the
 * image's zxy information.
 *
 * @param {Object} options
 * @param {Buffer} options.buffer - the buffer of the image to get pixel information from
 * @param {Number} options.imageSize - the size in pixels of the image. It must be square. Typical values are 256 or 512
 * @param {Number} options.lng - the longitude value
 * @param {Number} options.lat - the latitude value
 * @param {Number} options.x - the tile `x` coordinate
 * @param {Number} options.y - the tile `y` coordinate
 * @param {Number} options.z - the zoom level of the tile
 * @returns {Object} pixel information from a particular point on the image
 */
function getPixel(options) {
  // get tile x and y pixel locations
  var tileX = options.x * options.imageSize;
  var tileY = options.y * options.imageSize;

  // get pixel coordinates of the lat/lng values at specific zoom
  var pixelCoords = sm.px([options.lng, options.lat], options.z);

  // get relative pixel coordinates to the tile
  var x = pixelCoords[0] - tileX;
  var y = pixelCoords[1] - tileY;

  // generate new mapnik image object from the buffer
  var img = new mapnik.Image.fromBytesSync(options.buffer);

  // get pixel values at relative x/y positions for this image
  var values = img.getPixel(x, y, { get_color: true });

  return values;
}
```


CHANGELOG.md
# changelog

## 1.0.0

- Updated to mapnik@3.7.0
- Drops windows support

## 0.2.0

- Updated to mapnik@3.6.0

## 0.1.0

 - Updated mapnik to 3.5.0


test/test.js
var rtq = require('../');
var fs = require('fs');
var testPoints = require('./fixtures/test-coords');
var assert = require("assert");

function readTile(tile,callback) {
    var tilepath = 'test/fixtures/' + tile.z + '/' + tile.x + '/' + tile.y + '.png';
    fs.readFile(tilepath, function(err,data) {
        if (err && err.code === 'ENOENT') return callback(new Error('Tile does not exist'));
        if (err) return callback(err);

        return callback(null,data);
    });
}

describe('Load correct tiles', function() {
    it('should load correct tiles', function(done) {
        var correctTiles = [
            '{"z":16,"x":10642,"y":24989}',
            '{"z":16,"x":10643,"y":24989}',
            '{"z":16,"x":10642,"y":24990}',
            '{"z":16,"x":10643,"y":24990}',
            '{"z":16,"x":10642,"y":24991}'
        ];
        var options = {
            maxZoom: 16,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        rtq.loadTiles(testPoints.points, options, readTile, function(err,data) {
            for (var i = 0; i < data.length; i++) {
                assert.equal(JSON.stringify(data[i].zxy), correctTiles[i]);
            }
            done();
        });
    });

    it('should return empty string and empty: true if tile does not exist, but not fail if at least one tile has valid points', function(done) {
        var points = [[30, -150],[39.24282321740235, -121.53676986694335]];
        var options = {
            maxZoom: 16,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        var validResp = '{"zxy":{"z":16,"x":5461,"y":27038},"points":[[-150,30]],"pointIDs":[0],"data":"","empty":true}';
        rtq.loadTiles(points, options, readTile, function(err,data) {
            assert.equal(JSON.stringify(data[0]), validResp);
            done();
        });
    });

    it('should error if no tiles exist at any query location', function(done) {
        var points = [[30, -150], [32, -151]];
        var validErr = 'No tiles have any data';
        var options = {
            maxZoom: 16,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        rtq.loadTiles(points, options, readTile, function(err,data) {
            assert.equal(err.message, validErr);
            done();
        });
    });

    it('should return an error if tile is not square', function(done) {
        var points = [[0,0]];
        var zxy = {
            z: 0,
            x: 0,
            y: 0
        };
        var validErr = 'Invalid tile at 0/0/0';
        readTile(zxy, function(err, data) {
            rtq.getPixels(data, points, zxy, 256, [0], function(err, results) {
                assert.equal(err.message, validErr);
                done();
            });
        });
    });

    it('should return an error if loaded tile size does not match defined image size', function(done) {
        var zxy = {
            z: 16,
            x:10642,
            y:24989
        };
        var points = [[-121.53676986694335, 39.24282321740235]];
        var validErr = 'Tilesize 512 does not match image dimensions 256x256';
        readTile(zxy, function(err, data) {
            rtq.getPixels(data, points, zxy, 512, [0], function(err, results) {
                assert.equal(err.message, validErr);
                done();
            });
        });
    });

});

describe('Getting pixels', function() {
    it('should return the correct pixel XY', function(done) {
        var tileX = 10642;
        var tileY = 24989;
        var tileSize = 256;
        var absXY = [2724584, 6397316];
        var validResp = '{"x":232,"y":132}';
        var pxy = rtq.getPixelXY(tileX * tileSize, tileY * tileSize, absXY);
        assert.equal(JSON.stringify(pxy), validResp);
        done();
    });

    it('should return the correct pixel value from a valid query', function(done) {
        var zxy = {
            z: 16,
            x:10642,
            y:24989
        };
        var points = [[-121.53676986694335, 39.24282321740235]];
        var validResp = '[{"pixel":{"premultiplied":false,"a":255,"b":108,"g":72,"r":117},"latlng":{"lat":39.24282321740235,"lng":-121.53676986694335},"id":0}]';
        readTile(zxy, function(err, data) {
            rtq.getPixels(data, points, zxy, 256, [0], function(err, results) {
                assert.equal(JSON.stringify(results), validResp);
                done();
            });
        });
    });

    it('should return a null pixel value where tile does not exist', function(done) {
        var zxy = {
            z: 16,
            x:10642,
            y:24989
        };
        var points = [[-127.53676986694335, 32.24282321740235]];
        var validResp = '[{"pixel":null,"latlng":{"lat":32.24282321740235,"lng":-127.53676986694335},"id":0}]';
        readTile(zxy, function(err, data) {
            rtq.emptyPixelResponse(points, [0], function(err, results) {
                assert.equal(JSON.stringify(results), validResp);
                done();
            });
        });
    });

    it('should return error if the provided coordinates are not in the tile', function(done) {
        var zxy = {
            z: 16,
            x:10642,
            y:24989
        };
        var pixels = [[-120.5, 35.5]];
        var validErr = 'Coordinates are not in tile';
        readTile(zxy, function(err, data) {
            rtq.getPixels(data, pixels, zxy, 256, [0], function(err, results) {
                assert.equal(err.message, validErr);
                done();
            });
        });
    });

});

describe('Return the correct results from a query', function() {
    it('should work for multiple tile queries', function(done) {
        var validResp = '[{"pixel":{"premultiplied":false,"a":255,"b":108,"g":72,"r":117},"latlng":{"lat":39.24282321740235,"lng":-121.53676986694335},"id":0},{"pixel":{"premultiplied":false,"a":255,"b":116,"g":65,"r":162},"latlng":{"lat":39.241626684998266,"lng":-121.53685569763182},"id":1},{"pixel":{"premultiplied":false,"a":255,"b":119,"g":69,"r":170},"latlng":{"lat":39.241626684998266,"lng":-121.53565406799318},"id":2},{"pixel":{"premultiplied":false,"a":255,"b":118,"g":66,"r":171},"latlng":{"lat":39.24056308350469,"lng":-121.53642654418945},"id":3},{"pixel":{"premultiplied":false,"a":255,"b":115,"g":72,"r":145},"latlng":{"lat":39.239499465884755,"lng":-121.53582572937012},"id":4},{"pixel":{"premultiplied":false,"a":255,"b":98,"g":67,"r":104},"latlng":{"lat":39.23873498075964,"lng":-121.53672695159912},"id":5},{"pixel":{"premultiplied":false,"a":255,"b":107,"g":66,"r":133},"latlng":{"lat":39.23743866085578,"lng":-121.53651237487793},"id":6},{"pixel":{"premultiplied":false,"a":255,"b":115,"g":72,"r":152},"latlng":{"lat":39.236707392907185,"lng":-121.53779983520508},"id":7},{"pixel":{"premultiplied":false,"a":255,"b":101,"g":58,"r":125},"latlng":{"lat":39.23584315732298,"lng":-121.53668403625488},"id":8},{"pixel":{"premultiplied":false,"a":255,"b":107,"g":59,"r":161},"latlng":{"lat":39.23484594918998,"lng":-121.53762817382814},"id":9},{"pixel":{"premultiplied":false,"a":255,"b":108,"g":64,"r":163},"latlng":{"lat":39.23398,"lng":-121.53637},"id":10}]';
        var options = {
            maxZoom: 17,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        rtq.loadTiles(testPoints.points, options, readTile, function(err,data) {
            if (err) throw err;
            rtq.multiQuery(data,256,function(err,query) {
                assert.equal(JSON.stringify(query),validResp);
                done();
            });
        });
    });

    it('should work for multiple tile queries and select the correct zoom', function(done) {
        var validResp = '[{"pixel":{"premultiplied":false,"a":255,"b":103,"g":65,"r":137},"latlng":{"lat":39.24282321740235,"lng":-121.53676986694335},"id":0},{"pixel":{"premultiplied":false,"a":255,"b":111,"g":73,"r":150},"latlng":{"lat":39.241626684998266,"lng":-121.53685569763182},"id":1},{"pixel":{"premultiplied":false,"a":255,"b":117,"g":70,"r":164},"latlng":{"lat":39.241626684998266,"lng":-121.53565406799318},"id":2},{"pixel":{"premultiplied":false,"a":255,"b":112,"g":67,"r":154},"latlng":{"lat":39.24056308350469,"lng":-121.53642654418945},"id":3},{"pixel":{"premultiplied":false,"a":255,"b":112,"g":73,"r":147},"latlng":{"lat":39.239499465884755,"lng":-121.53582572937012},"id":4},{"pixel":{"premultiplied":false,"a":255,"b":102,"g":63,"r":137},"latlng":{"lat":39.23873498075964,"lng":-121.53672695159912},"id":5},{"pixel":{"premultiplied":false,"a":255,"b":114,"g":78,"r":138},"latlng":{"lat":39.23743866085578,"lng":-121.53651237487793},"id":6},{"pixel":{"premultiplied":false,"a":255,"b":118,"g":72,"r":155},"latlng":{"lat":39.236707392907185,"lng":-121.53779983520508},"id":7},{"pixel":{"premultiplied":false,"a":255,"b":106,"g":60,"r":141},"latlng":{"lat":39.23584315732298,"lng":-121.53668403625488},"id":8},{"pixel":{"premultiplied":false,"a":255,"b":107,"g":66,"r":143},"latlng":{"lat":39.23484594918998,"lng":-121.53762817382814},"id":9},{"pixel":{"premultiplied":false,"a":255,"b":121,"g":95,"r":148},"latlng":{"lat":39.23398,"lng":-121.53637},"id":10}]';
        var options = {
            maxZoom: 17,
            minZoom: 9,
            tileSize: 256
        }
        rtq.loadTiles(testPoints.points, options, readTile, function(err,data) {
            if (err) throw err;
            rtq.multiQuery(data,256,function(err,query) {
                assert.equal(JSON.stringify(query),validResp);
                done();
            });
        });
    });


    it('should work for single-point queries', function(done) {
        var point = [[39.24282321740235,-121.53676986694335]];
        var validResp = '[{"pixel":{"premultiplied":false,"a":255,"b":108,"g":72,"r":117},"latlng":{"lat":39.24282321740235,"lng":-121.53676986694335},"id":0}]';
        var options = {
            maxZoom: 17,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        rtq.loadTiles(point, options, readTile, function(err,data) {
            if (err) throw err;
            rtq.multiQuery(data,256,function(err,query) {
                assert.equal(JSON.stringify(query),validResp)
                done();
            });
        });
    });

    it('should return null for points in a non-existent tile, but not fail if at least one tile has valid points', function(done) {
        var validResp = '[{"pixel":null,"latlng":{"lat":33.24282321740235,"lng":-120.53676986694335},"id":0},{"pixel":null,"latlng":{"lat":33.241626684998266,"lng":-120.53685569763182},"id":1},{"pixel":null,"latlng":{"lat":33.241626684998266,"lng":-121.53565406799318},"id":2},{"pixel":{"premultiplied":false,"a":255,"b":118,"g":66,"r":171},"latlng":{"lat":39.24056308350469,"lng":-121.53642654418945},"id":3},{"pixel":{"premultiplied":false,"a":255,"b":115,"g":72,"r":145},"latlng":{"lat":39.239499465884755,"lng":-121.53582572937012},"id":4},{"pixel":{"premultiplied":false,"a":255,"b":98,"g":67,"r":104},"latlng":{"lat":39.23873498075964,"lng":-121.53672695159912},"id":5},{"pixel":{"premultiplied":false,"a":255,"b":107,"g":66,"r":133},"latlng":{"lat":39.23743866085578,"lng":-121.53651237487793},"id":6},{"pixel":{"premultiplied":false,"a":255,"b":115,"g":72,"r":152},"latlng":{"lat":39.236707392907185,"lng":-121.53779983520508},"id":7},{"pixel":{"premultiplied":false,"a":255,"b":101,"g":58,"r":125},"latlng":{"lat":39.23584315732298,"lng":-121.53668403625488},"id":8},{"pixel":null,"latlng":{"lat":40.23484594918998,"lng":-111.53762817382814},"id":9},{"pixel":null,"latlng":{"lat":40.23398,"lng":-111.53637},"id":10}]';
        var options = {
            maxZoom: 16,
            minZoom: 15,
            zoom: 16,
            tileSize: 256
        }
        rtq.loadTiles(testPoints.pointsSomeNull, options, readTile, function(err,data) {
            if (err) throw err;
            rtq.multiQuery(data,256,function(err,query) {
                assert.equal(JSON.stringify(query),validResp);;
                done();
            });
        });
    });
});

describe('Utility functions should work as intended', function() {
    it('should select the correct zoom level where tile pixel spacing ~= bounding box density', function(done) {
        var extent = [39.23398,-121.53779983520508,39.24282321740235,-121.53565406799318];
        var smExtent = {
            lowerLeft: [-13529525.989789838, 4755242.832925649],
            upperRight: [ -13529287.124076448, 4756513.841501413]
        }
        var queryLength = 100;
        var tileSize = 256;
        var expectedZoom = 14;
        var estimatedZoom = rtq.estimatePixelSnap(extent, smExtent, queryLength, tileSize);
        assert.equal(expectedZoom, estimatedZoom);
        done();
    });

    it('should correctly estimate the correct zoom based on an length > 1 array of input points', function(done) {
        var maxZoom = 17;
        var minZoom = 10;
        var tileSize = 256;
        var expectedZoom = 11;
        var estimatedZoom = rtq.estimateZoom(testPoints.points, minZoom, maxZoom, tileSize);
        assert.equal(expectedZoom, estimatedZoom);
        done();
    });

    it('should correctly estimate the correct zoom ... ^ ...., but not lower than the minZoom', function(done) {
        var maxZoom = 17;
        var minZoom = 12;
        var tileSize = 256;
        var expectedZoom = 12;
        var estimatedZoom = rtq.estimateZoom(testPoints.points, minZoom, maxZoom, tileSize);
        assert.equal(expectedZoom, estimatedZoom);
        done();
    });

    it('should correctly estimate the correct zoom ... ^ ...., but not higher than the maxZoom', function(done) {
        var maxZoom = 8;
        var minZoom = 1;
        var tileSize = 256;
        var expectedZoom = 8;
        var estimatedZoom = rtq.estimateZoom(testPoints.points, minZoom, maxZoom, tileSize);
        assert.equal(expectedZoom, estimatedZoom);
        done();
    });

    it('should select the maxZoom for a query of length 1', function(done) {
        var maxZoom = 17;
        var minZoom = 12;
        var tileSize = 256;
        var expectedZoom = 17;
        var estimatedZoom = rtq.estimateZoom([testPoints.points[0]], minZoom, maxZoom, tileSize);
        assert.equal(expectedZoom, estimatedZoom);
        done();
    });
});

test/fixtures/test-coords.js
var points = [
    [
        39.24282321740235, 
        -121.53676986694335
    ], 
    [
        39.241626684998266, 
        -121.53685569763182
    ], 
    [
        39.241626684998266, 
        -121.53565406799318
    ], 
    [
        39.24056308350469, 
        -121.53642654418945
    ], 
    [
        39.239499465884755, 
        -121.53582572937012
    ], 
    [
        39.23873498075964, 
        -121.53672695159912
    ], 
    [
        39.23743866085578, 
        -121.53651237487793
    ], 
    [
        39.236707392907185, 
        -121.53779983520508
    ], 
    [
        39.23584315732298, 
        -121.53668403625488
    ], 
    [
        39.23484594918998, 
        -121.53762817382814
    ], 
    [
        39.23398, 
        -121.53637
    ]
];

var pointsSomeNull = [
    [
        33.24282321740235, 
        -120.53676986694335
    ], 
    [
        33.241626684998266, 
        -120.53685569763182
    ], 
    [
        33.241626684998266, 
        -121.53565406799318
    ], 
    [
        39.24056308350469, 
        -121.53642654418945
    ], 
    [
        39.239499465884755, 
        -121.53582572937012
    ], 
    [
        39.23873498075964, 
        -121.53672695159912
    ], 
    [
        39.23743866085578, 
        -121.53651237487793
    ], 
    [
        39.236707392907185, 
        -121.53779983520508
    ], 
    [
        39.23584315732298, 
        -121.53668403625488
    ], 
    [
        40.23484594918998, 
        -111.53762817382814
    ], 
    [
        40.23398, 
        -111.53637
    ]
];

module.exports = {
    points: points,
    pointsSomeNull: pointsSomeNull
}

index.js
var mapnik = require('mapnik');
var sphericalmercator = require('sphericalmercator');
var sm;
var async = require('queue-async');

/**
 * Get the pixel value information for a given array of lat/long coordinate pairs
 *
 * @param {Buffer} imageBuffer
 * @param {Array} coords - array of x/y point arrays
 * @param {Object} zxy - zxy object in the form of `{z: z, x: x, y: y}`
 * @param {number} tileSize - pixel size of the tile (all tiles are perfect squares)
 * @param {Array} ids - array of ids to represent the response for each coordinate in `coords`
 * @param {Function} callback - `function(err, results)`
 *
 * @example
 * var rtq = require('@mapbox/raster-tile-query');
 * var fs = require('fs');
 * var img = fs.readFileSync('./path/to/tile.png');
 * var zxy = {
 *    z: 16,
 *    x:10642,
 *    y:24989
 * };
 * var points = [[-121, 39]];
 * rtq.getPixels(img, points, zxy, 256, [0], function(err, results) {
 *   console.log(results);
 *   // [
 *   //   {
 *   //     "pixel": {
 *   //       "premultiplied": false,
 *   //       "a":255,
 *   //       "b":108,
 *   //       "g":72,
 *   //       "r":117
 *   //     },
 *   //     "latlng": {
 *   //       "lat":39,
 *   //       "lng":-121
 *   //     },
 *   //     "id":0
 *   //   }
 *   // ]
 * });
 */
function getPixels(imageBuffer, coords, zxy, tileSize, ids, callback) {

    var image = mapnik.Image.fromBytesSync(imageBuffer);
    var iWidth = image.width();
    var iHeight = image.height();

    if (iWidth !== iHeight) {
        return callback(new Error('Invalid tile at ' + zxy.z + '/'+ zxy.x + '/'+ zxy.y));
    } else if (iWidth !== tileSize) {
        return callback(new Error('Tilesize ' + tileSize + ' does not match image dimensions ' + iWidth + 'x' + iHeight));
    }

    var tileX = zxy.x * tileSize;
    var tileY = zxy.y * tileSize;
    var output = [];

    for (var i = 0; i < coords.length; i ++) {
        var pCoords = sm.px(coords[i], zxy.z);
        var xy = getPixelXY(tileX, tileY, pCoords);
        if (xy.x >= tileSize || xy.y >= tileSize) return callback(new Error('Coordinates are not in tile'));
        var queryResult = {
            pixel: image.getPixel(xy.x, xy.y, {get_color:true}),
            latlng: {
                lat: coords[i][1],
                lng: coords[i][0]
            },
            id: ids[i]
        };
        output.push(queryResult);

    }
    return callback(null,output);
}

function emptyPixelResponse(coords, ids, callback) {
    var output = [];
    for (var i = 0; i < coords.length; i ++) {
        var queryResult = {
            pixel: null,
            latlng: {
                lat: coords[i][1],
                lng: coords[i][0]
            },
            id: ids[i]
        };
        output.push(queryResult);
    }
    return callback(null,output);
}

function sortBy(sortField) {
    return function sortCallback(a, b) {
        var ad = a[sortField] || 0;
        var bd = b[sortField] || 0;
        return ad < bd ? -1 : ad > bd ? 1 : 0;
    };
}

/**
 *
 */
function getPixelXY(tileX, tileY, pixel) {
    return {
        x: pixel[0] - tileX,
        y: pixel[1] - tileY
    };
}

function buildQuery(points, zoom) {
    var queryObject = {}, output = [];
    for (var i = 0; i < points.length; i++) {
        var xyz = sm.xyz([points[i][1], points[i][0], points[i][1], points[i][0]], zoom);
        var tileName = zoom + '/' + xyz.minX + '/' + xyz.minY;
        if (queryObject[tileName] === undefined) {
            queryObject[tileName] = {
                zxy: {
                    z: zoom,
                    x: xyz.minX,
                    y: xyz.minY
                },
                points: [
                    [points[i][1], points[i][0]]
                ],
                pointIDs: [i]
            };
            output.push(queryObject[tileName]);
        } else {
            queryObject[tileName].points.push([points[i][1], points[i][0]]);
            queryObject[tileName].pointIDs.push(i);
        }
    }
    return output;
}

function getExtent(points) {
    var bounds = [points[0][0], points[0][1], points[0][0],points[0][1]];
    return points.reduce(function(a, b) {
        if (bounds[0] > b[0]) {
            bounds[0] = b[0];
        } else if (bounds[2] < b[0]){
            bounds[2] = b[0];
        }
        if (bounds[1] > b[1]) {
            bounds[1] = b[1];
        } else if (bounds[3] < b[1]) {
            bounds[3] = b[1];
        }
        return bounds;
    });
}

function estimatePixelSnap(extent, smExtent, queryLength, tileSize) {
    smExtent.xRange = smExtent.upperRight[0] - smExtent.lowerLeft[0];
    smExtent.yRange = smExtent.upperRight[1] - smExtent.lowerLeft[1];
    var pRatio;
    if (smExtent.xRange > smExtent.yRange) {
        pRatio = ((extent[3] - extent[1]) / Math.ceil((smExtent.xRange / (smExtent.xRange + smExtent.yRange)) * queryLength * 2) * tileSize);
        return Math.ceil((Math.log(1 / (extent[3] - extent[1])) + Math.log(360)) / Math.log(2));
    } else {
        pRatio = ((extent[2] - extent[0]) / Math.ceil((smExtent.yRange / (smExtent.xRange + smExtent.yRange)) * queryLength * 2) * tileSize);
        return Math.ceil((Math.log(1 / (pRatio)) + Math.log(170.10225756)) / Math.log(2));
    }
}

function estimateZoom(queryPoints, minZoom, maxZoom, tileSize) {
    if (queryPoints.length === 1) {
        return maxZoom;
    } else {
        if (!sm) {
            sm = new sphericalmercator({
                size: tileSize
            });
        }
        var extent = getExtent(queryPoints);
        var smExtent = {
            lowerLeft: sm.forward([extent[1], extent[0]]),
            upperRight: sm.forward([extent[3], extent[2]])
        };
        var estZ = estimatePixelSnap(extent, smExtent, queryPoints.length, tileSize);

        if (estZ > maxZoom) {
            estZ = maxZoom;
        } else if (estZ < minZoom) {
            estZ = minZoom;
        }
        return estZ;
    }
}

function loadTiles(queryPoints, options, loadFunction, callback) {
    if (!queryPoints[0].length) return callback(new Error('Invalid query points'));

    if (options.maxZoom === undefined) return callback(new Error('Max zoom must be specified'));

    if (options.minZoom === undefined) return callback(new Error('Min zoom must be specified'));

    var minZoom = options.minZoom;
    var maxZoom = options.maxZoom;
    var tileSize = options.tileSize || 256;

    if (!sm) {
        sm = new sphericalmercator({
            size: tileSize
        });
    }

    var zoom = options.zoom !== undefined ? options.zoom : estimateZoom(queryPoints, minZoom, maxZoom, tileSize);

    var nullcount = 0;

    function loadTileAsync(tileObj, loadFunction, callback) {
        loadFunction(tileObj.zxy, function(err, data) {
            if (err && err.message === 'Tile does not exist') {
                tileObj.data = '';
                tileObj.empty = true;
                nullcount++;
                if (nullcount === tileQuerier.length) {
                    return callback(new Error('No tiles have any data'));
                } else {
                    return callback(null, tileObj);
                }
            }
            if (err) return callback(err);
            tileObj.data = data;
            return callback(null, tileObj);
        });
    }

    var tileQuerier = buildQuery(queryPoints, zoom);
    var loadQueue = new async();
    for (var i = 0; i < tileQuerier.length; i++) {
        loadQueue.defer(loadTileAsync, tileQuerier[i], loadFunction);
    }

    loadQueue.awaitAll(callback);
}

function multiQuery(tileQuerier,imageSize,callback) {

    function queriesDone(err, queries) {
        if (err) return callback(err);
        var dataOutput = [];
        dataOutput = dataOutput.concat.apply(dataOutput, queries);
        dataOutput.sort(sortBy('id'));
        return callback(null, dataOutput);
    }

    var queryQueue = new async();

    for (var i = 0; i<tileQuerier.length; i++) {
        if (tileQuerier[i].empty) {
            queryQueue.defer(emptyPixelResponse, tileQuerier[i].points,tileQuerier[i].pointIDs);
        } else {
            queryQueue.defer(getPixels, tileQuerier[i].data, tileQuerier[i].points, tileQuerier[i].zxy, imageSize, tileQuerier[i].pointIDs);
        }
    }

    queryQueue.awaitAll(queriesDone);
}

module.exports = {
    getPixels: getPixels,
    buildQuery: buildQuery,
    loadTiles: loadTiles,
    multiQuery: multiQuery,
    getPixelXY: getPixelXY,
    emptyPixelResponse: emptyPixelResponse,
    estimateZoom: estimateZoom,
    estimatePixelSnap: estimatePixelSnap
};
